integration/spark/src/main/scala/org/apache/spark/sql/hive/DistributionUtil.scala

```diff
@@ -101,47 +101,107 @@ object DistributionUtil {
+      // this case will come only if dynamic allocation is true
+      var totalExecutorsToBeRequested = nodesOfData
+      // If total number of blocks are greater than the nodes identified then ensure
+      // that the configured number of max executors can be opened based on the difference of
+      // block list size and nodes identified
+      if (blockList.size > nodesOfData) {
+        // e.g 1. blockList size = 40, nodesOfData = 3, confExecutors = 6, then all executors
+        // need to be opened
+        // 2. blockList size = 4, nodesOfData = 3, confExecutors = 6, then
+        // total 4 executors need to be opened
+        val extraExecutorsToBeRequested = blockList.size - nodesOfData
+        if (extraExecutorsToBeRequested > confExecutors) {
+          totalExecutorsToBeRequested = confExecutors
+        } else {
+          totalExecutorsToBeRequested = nodesOfData + extraExecutorsToBeRequested
+        }
       }
+      LOGGER.info("Total executors requested: " + totalExecutorsToBeRequested)
+      totalExecutorsToBeRequested
+    } else {
+      nodesOfData
     }
 
-    val confExecutors = if (null != confExecutorsTemp) {
-      confExecutorsTemp.toInt
+    val startTime = System.currentTimeMillis();
+    if (sparkContext.getConf.getBoolean("spark.dynamicAllocation.enabled", false)) {
+      // this case will come only if dynamic allocation is true
+      CarbonContext
+        .ensureExecutors(sparkContext, requiredExecutors, blockList.size, Map.empty)
     } else {
-      1
+      CarbonContext.ensureExecutors(sparkContext, requiredExecutors)
     }
-    val requiredExecutors = if (nodeMapping.size > confExecutors) {
-      confExecutors
+    getDistinctNodesList(sparkContext, requiredExecutors, startTime)
+  }
+
+  /**
+   * This method will return the configured executors
+   *
+   * @param sparkContext
+   * @return
+   */
+  private def getConfiguredExecutors(sparkContext: SparkContext): Int = {
+    var confExecutors: Int = 0
+    if (sparkContext.getConf.getBoolean("spark.dynamicAllocation.enabled", false)) {
+      // default value for spark.dynamicAllocation.maxExecutors is infinity
+      confExecutors = sparkContext.getConf.getInt("spark.dynamicAllocation.maxExecutors", 1)
     } else {
-      nodeMapping.size()
+      // default value for spark.executor.instances is 2
+      confExecutors = sparkContext.getConf.getInt("spark.executor.instances", 1)
     }
+    confExecutors
+  }
 
-    val startTime = System.currentTimeMillis()
-    CarbonContext.ensureExecutors(sparkContext, requiredExecutors)
+  /**
+   * This method will return the distinct nodes list
+   *
+   * @param sparkContext
+   * @param requiredExecutors
+   * @param startTime
+   * @return
+   */
+  private def getDistinctNodesList(sparkContext: SparkContext,
+      requiredExecutors: Int,
+      startTime: Long): Seq[String] = {
     var nodes = DistributionUtil.getNodeList(sparkContext)
-    var maxTimes = 30
+    var maxTimes = 10;
```

Comment:
Earlier waiting time to request the executors was 15 secs which is too long. I have reduced it to 5 secs. I have tested a case of starting 70 executors and in 5 sec waiting time it was able to start all 70 executors. Even if it does not start all 70 executors for the first query, second query onward we can get the benefit of it.